商品編集

K-Hir-pk18 opened the pull request:
WHAT
・商品編集機能

WHY
・商品情報の変更に対応するため

K-Hir-pk18 committed:
商品編集

WHAT
・商品編集機能

WHY
・商品情報の変更に対応するため

K-Hir-pk18 commented:
出品ログインユーザーボタン表示
https://gyazo.com/378c85cd1f15a6c6c233633e90c233c2

非出品ログインユーザーURL直接入力時
https://gyazo.com/a082e97e5fd85db6870198463b08610d

非ログインユーザーURL直接入力時
https://gyazo.com/3cfa254f7a941f36f845f83326e424e3

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

kata-myu reviewed (commented):
こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、追加で下記キャプチャ画像をご提出いただけますでしょうか。
```
・商品編集に成功する動画
・商品編集に失敗し、エラーが表示される動画
・何も編集せずに更新をしても画像無しの商品にならない動画
・商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
```

お手数ですが、よろしくお願いいたします。

K-Hir-pk18 commented:
出品ログインユーザーボタン表示
https://gyazo.com/378c85cd1f15a6c6c233633e90c233c2

非出品ログインユーザーURL直接入力時
https://gyazo.com/a082e97e5fd85db6870198463b08610d

非ログインユーザーURL直接入力時
https://gyazo.com/3cfa254f7a941f36f845f83326e424e3

・商品編集に成功
https://gyazo.com/e10f461593555d62fbde164ac0a3f0bb
https://gyazo.com/fb11985e746e7997ee1b2078aece44d8

・商品編集に失敗し、エラーが表示
https://gyazo.com/372a7700bd6b4a348b7202e7f21b213d

・何も編集せずに更新をしても画像無しの商品にならない
https://gyazo.com/63129d640c710c1cba00ca75020106ec

・商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される
https://gyazo.com/a5acb55d71407c2e7d3dba951ad4bf1d

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

s-aka-moto commented on app/controllers/products_controller.rb:
```diff
@@ -24,31 +23,34 @@ def create
+    end
+  end
+
+  def update
```
updateアクションにも「出品者でない場合」の挙動を記述しましょう！
(edit同様、indexアクションにリダイレクトする場合はメソッドを定義し、before_actionでeditとupdateに適用すると記述がスッキリしますね。)
```
before_action : 出品者でない場合はindexアクションにリダイレクトするメソッド[:edit,:update]
...
private 
def 出品者でない場合はindexアクションにリダイレクトするメソッド
if current_user.id != @product.user_id
      redirect_to action: :index
    end
end
```
)

検証画面でフォームを作成し、情報を書き換えることができてしまうため、悪意のあるユーザーからの攻撃を避けるためです。

s-aka-moto commented on app/controllers/products_controller.rb:
```diff
@@ -24,31 +23,34 @@ def create
+
+  def update
+    @product.update(product_params)
+    unless @product.save
```
こちらはご提案となりますが、updateに成功した場合は詳細画面に遷移する、という条件も加えてみてはいかがでしょうか？(updateに成功したことがわかりやすいため。)

K-Hir-pk18 replied:
s-aka-moto様

ご対応ありがとうございます。

とにかく目の前の課題をこなすだけで精一杯の自分にとって
良い勉強になるアドバイスです。

updateもpath?を入力すると割り込めるのですね！

CoC（今復習しました…）だと開発を進めやすくできる反面、
こういった（攻撃者にとって標的箇所が容易に予測できる）リスクもあるのは
まったく気にしていなかったです。（まだ無縁の世界だと思っている状態だったので）

またためになるアドバイスをいただけましたらば幸いです。

s-aka-moto reviewed (commented):
レビュー依頼gyazoのご送付ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします！

K-Hir-pk18 committed:
商品編集機能　レビュー対応２

WHAT
・product_controller
 　　→　private_method confirm_user
　　　　 before_action confirm_user
 　　　　updateアクション

WHY
・悪意あるユーザーの攻撃を回避するため
・ユーザビリティーの向上

K-Hir-pk18 commented:
　ご担当者様


よろしくお願いいたします。

chirikuzu commented on app/controllers/products_controller.rb:
```diff
@@ -24,31 +24,39 @@ def create
+  end
+
+  def update
+    @product.update(product_params)
+    unless @product.save
+      render :edit
+    else
+      redirect_to action: :show
+    end
```
こちら、修正されてより良いコードとなりましたね。
今回の記述でも、保存に失敗した場合はeditのビューファイルの呼び出し、保存に成功した場合はshowアクションにリダイレクトさせる記述としては問題ありません。
そのため以下は参考程度にご認識いただければと存じます。

今回の場合であれば`@product.update(product_params)`は、`save`と同様にtrue(更新に成功)あるいはfalse(更新に失敗)という結果を返します。
そのため、unless文を用いて`保存に失敗した場合`という条件を記述しなくても、下記のようにエラーハンドリングを行うことも可能です。
```ruby
if @product.update(item_params) # 更新に成功した場合
  redirect_to action: :show
else # 更新に失敗した場合
  render :edit
end
```
unless文には一見するとどういう場合を想定しているのかが第三者に伝わりづらくなってしまう場合もございますので、こちらコメントさせていただきました。
ご参考になれば幸いです。

K-Hir-pk18 replied:
chirikuzu様

レビューいただきましてありがとうございます。
コメントの方も勉強になります。

updateメソッドは最初、unlessで記述していたのでその名残でした。

ただ仰る通り、可読性も考えifの方が良いと思いますので修正します。
（意地を張ってunlessのままでも良いのですが、無意味なのでｗ）

またレビューなどでお世話になる機会がありましたら
有益なことを色々と教えていただけると幸いです。
よろしくお願いいたします。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れ様でした。

1点だけコメントさせていただいておりますが、そちらについてはご参考程度にご認識いただければ幸いです。

マージ、プルを実行後、次の実装に進まれてください。
引き続き実装頑張っていきましょう。
